File: Codes/2411-smallest-subarrays-with-maximum-bitwise-or.cpp
/*
 * @Author: LetMeFly
 * @Date: 2025-07-29 23:42:57
 * @LastEditors: LetMeFly.xyz
 * @LastEditTime: 2025-07-30 00:00:01
 */
#if defined(_WIN32) || defined(__APPLE__)
#include "_[1,2]toVector.h"
#endif

class Solution {
public:
    vector<int> smallestSubarrays(vector<int>& nums) {
        vector<int> ans(nums.size());
        vector<int> lastPos(31);
        for (int i = nums.size() - 1; i >= 0; i--) {
            int last = i;
            for (int j = 0; j < 31; j++) {
                if (nums[i] >> j & 1) {
                    lastPos[j] = i;
                } else {
                    last = max(last, lastPos[j]);
                }
            }
            ans[i] = max(ans[i], last - i + 1);
        }
        return ans;
    }
};


#if defined(_WIN32) || defined(__APPLE__)
/*
 [1,0,2,1,3]
*/
int main() {
    string s;
    while (cin >> s) {
        vector<int> v = stringToVector(s);
        Solution sol;
        debug(sol.smallestSubarrays(v));
    }
    return 0;
}
#endif

File: Codes/2411-smallest-subarrays-with-maximum-bitwise-or.go
/*
 * @Author: LetMeFly
 * @Date: 2025-07-29 23:42:57
 * @LastEditors: LetMeFly.xyz
 * @LastEditTime: 2025-07-30 10:15:13
 */
package main

func smallestSubarrays(nums []int) []int {
    ans := make([]int, len(nums))
    lastPos := make([]int, 31)
    for i := len(nums) - 1; i >= 0; i-- {
        last := i
        for j := 0; j < 31; j++ {
            if nums[i] >> j & 1 == 1 {
                lastPos[j] = i
            } else {
                last = max(last, lastPos[j])
            }
        }
        ans[i] = max(ans[i], last - i + 1)
    }
    return ans
}